File: front/lib/domain/model/owner_invitation.dart
import 'package:amap_en_ligne/domain/model/invitation_status.dart';
import 'package:freezed_annotation/freezed_annotation.dart';

part 'owner_invitation.freezed.dart';
part 'owner_invitation.g.dart';

@freezed
abstract class OwnerInvitation with _$OwnerInvitation {
  const factory OwnerInvitation({
    @JsonKey(name: 'invitation_id') required String invitationId,
    @JsonKey(name: 'first_name') required String firstName,
    @JsonKey(name: 'last_name') required String lastName,
    required String email,
    required InvitationStatus status,
    @JsonKey(name: 'submitted_at') required String submittedAt,
    @JsonKey(name: 'resend_requested_at') String? resendRequestedAt,
    @JsonKey(name: 'activated_at') String? activatedAt,
  }) = _OwnerInvitation;

  factory OwnerInvitation.fromJson(Map<String, Object?> json) =>
      _$OwnerInvitationFromJson(json);
}

File: back/persistence/model/src/main/kotlin/OwnerInvitation.kt
@file:OptIn(ExperimentalTime::class)

package persistence.model

import id.Id
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlin.time.ExperimentalTime
import kotlin.time.Instant

@Serializable
enum class OwnerInvitationStatus {
    PENDING_ACTIVATION,
    ACTIVATED,
    CANCELLED,
}

@Serializable
data class OwnerInvitation(
    @SerialName("invitation_id") val invitationId: Id<OwnerInvitation>,
    @SerialName("first_name") val firstName: String,
    @SerialName("last_name") val lastName: String,
    val email: String,
    val status: OwnerInvitationStatus,
    @SerialName("submitted_at") val submittedAt: Instant,
    @SerialName("resend_requested_at") val resendRequestedAt: Instant? = null,
    @SerialName("activated_at") val activatedAt: Instant? = null,
)

@Serializable
data class CreateOwnerInvitationBody(
    @SerialName("first_name") val firstName: String,
    @SerialName("last_name") val lastName: String,
    val email: String,
)

@Serializable
data class OwnerInvitationCreated(
    @SerialName("invitation_id") val invitationId: Id<OwnerInvitation>,
    val status: OwnerInvitationStatus,
)
